### ISSUE TITLE
Support for custom cross-validation strategies, e.g., forward chaining for temporally ordered tabular data

### ISSUE THREAD

**Issue by willadamskeane** [autogluon/autogluon#4492]
Hi AutoGluon team,

I'm using TabularPredictor for temporally ordered tabular data and need more flexible cross-validation options. The current CV methods and "groups" parameter don't suffice for many time-based scenarios.

Standard k-fold CV can cause data leakage with ordered data.
As far as I can tell, I'm unable to do the following using the "groups" parameter:

fold 1: training [month 1], test [month 2]
fold 2: training [months 1-2], test [month 3]
fold 3: training [months 1-3], test [month 4]
...and so on

Proposed solution:
Add a custom_cv_strategy parameter to TabularPredictor, allowing custom splitting functions or pre-implemented strategies like forward chaining.

Thank you for considering!

**Comment by Innixma** [comment]
Thanks for creating an issue! This would indeed be nice.

The tricky part comes with making this work during stack ensembling, and with how we perform model evaluation of bags.

## Stack Ensembling

For stacking, I think this simply wouldn't work unless we got **very** creative. This would be a major endeavor and I'd prefer to only tackle this after the initial implementation that works for non-stacked models.

## Model Evaluation

For bagged models, we compute the metric by merging the predictions of all fold models together and computing the metric. This wouldn't be correct if we perform the kind of cross-validation strategy you propose, instead we should compute the score of each fold model separately and then average.

This would require some additional work, and might require fundamentally changing the way we score bagged models to always average scores of fold models rather than combine preds, but this has the downside of not accurately tracking the performance of the bag when using repeated k-fold.

## Timeline

I've very tentatively marked this for the 1.2 release, but wouldn't be surprised if it is pushed to 1.3.
If you are interested in getting the feature earlier, you can give a try at implementing it.

The current CV Splitting code is here: https://github.com/autogluon/autogluon/blob/master/core/src/autogluon/core/utils/utils.py#L37-L604

One idea for implementation is the user could specify groups as a np.ndarray / pd.DataFrame where each column is a fold, and each value is `0` = Train, `1` = Test, `2` = Don't use.

## Example of 3 folds akin to 

> fold 1: training [month 1], test [month 2]
fold 2: training [months 1-2], test [month 3]
fold 3: training [months 1-3], test [month 4]

```
0 0 0
1 0 0 
2 1 0 
2 2 1
```

### Code Suggestion

An easy way to implement this as a prototype would be to get it working with the bagged model without worrying about the TabularPredictor integration. You can follow this tutorial on fitting bagged model outside of TabularPredictor: https://auto.gluon.ai/stable/tutorials/tabular/advanced/tabular-custom-model.html#training-a-bagged-custom-model-without-tabularpredictor

Then just pass `groups` into the `model.fit` call in whatever format you like, and check that the models are fit with the intended splits.

**Comment by lmsanch** [comment]
> Hi AutoGluon team,
> 
> I'm using TabularPredictor for temporally ordered tabular data and need more flexible cross-validation options. The current CV methods and "groups" parameter don't suffice for many time-based scenarios.
> 
> Standard k-fold CV can cause data leakage with ordered data. As far as I can tell, I'm unable to do the following using the "groups" parameter:
> 
> fold 1: training [month 1], test [month 2] fold 2: training [months 1-2], test [month 3] fold 3: training [months 1-3], test [month 4] ...and so on
> 
> Proposed solution: Add a custom_cv_strategy parameter to TabularPredictor, allowing custom splitting functions or pre-implemented strategies like forward chaining.
> 
> Let me know if I've overlooked something, and this is actually currently possible, or if there is a workaround.
> 
> Otherwise, thank you for considering!

@willadamskeane I've decided to switch to AutoGluon, moving away from my previous work involving stacking XGBoost and CatBoost models. However, since my work primarily involves financial data and quantitative analysis, maintaining the data’s sequential order is crucial, and use my own CV partitions. As it stands, AutoGluon isn't quite usable for my needs. Were you able to implement the proposed solution? I'd be interested in collaborating on this if you're open to it.

**Comment by falrrema** [comment]
Hi, this would be a killer feature in autogluon for me. I have experimented with autogluon for numerai competition the past year. I have generated interesting models, but are generally fragile. I need to do a lot post-processing on the predictions. I believe models would be significantly more robust if I am able to generate a walk-forward CV (with purging ideally) and if the stacking / ensembling also takes in consideration this aspect. If we could provide the CV partitions would be even better.

**Comment by ijmiller2** [comment]
+1 for custom cross validation strategies. I've been working a lot recently with chemical property modeling and splitting based on chemical structural similarity is crucial to prevent data leakage/overfitting in this domain. AutoGluon is awesome and would be much more practically useful to the chemistry/biology domains if we could specify custom CV splits.

**Comment by Innixma** [comment]
This is now supported. Thanks for the clear write-up — the exact schedule you described is what `validation_structure` produces.

## @willadamskeane — your scenario

```python
predictor = TabularPredictor(label="y").fit(
    train_data,
    num_bag_folds=5,
    validation_structure={"time_on": "month", "temporal_forward_only": True},
)
```

On 6 months of data with `num_bag_folds=5`:

```
model=LightGBM_BAG_L1 folds=5
months with an out-of-fold prediction: [2, 3, 4, 5, 6]
months left unvalidated (train-only):  [1]
```

Fold *i* trains on the blocks before block *i+1* and validates on block *i+1* — train [m1] → test [m2], train [m1-2] → test [m3], and so on. `time_on` accepts a datetime or numeric column; blocks are contiguous in time, and the block count follows `num_bag_folds`.

It matches `sklearn.model_selection.TimeSeriesSplit` fold for fold:

```
TimeSeriesSplit(n_splits=5)                      temporal_forward_only, num_folds=5
  fold 0: train t=[0..9]  val t=[10..19]           fold 0: train t=[0..9]  val t=[10..19]
  fold 1: train t=[0..19] val t=[20..29]           fold 1: train t=[0..19] val t=[20..29]
  fold 2: train t=[0..29] val t=[30..39]           fold 2: train t=[0..29] val t=[30..39]
  fold 3: train t=[0..39] val t=[40..49]           fold 3: train t=[0..39] val t=[40..49]
  fold 4: train t=[0..49] val t=[50..59]           fold 4: train t=[0..49] val t=[50..59]
```

The default (`temporal_forward_only=False`) is leave-one-block-out: contiguous time blocks, every block validated once, but each fold trains partly on the future. Use `temporal_forward_only=True` when the lookahead matters, at the cost of the earliest block never being validated and early folds seeing less data.

## @Innixma — the two blockers you raised

**Model evaluation.** Handled, and in the direction you suggested. Forward-chaining leaves the earliest block unvalidated, so those rows get no out-of-fold prediction; `score_with_oof` masks on fold coverage, so the bagged score is computed over the covered rows only. Uncovered rows are reported as missing rather than as a prediction of 0. The weighted ensemble is scored on the same row set, so it stays comparable with the models it is ranked against — with a single base model at weight 1.0 the two scores coincide exactly.

**Stack ensembling.** Not attempted. Combining `temporal_forward_only` with `num_stack_levels > 0` raises rather than silently producing a leaky stack, which matches your preference to land the non-stacked case first.

## @ijmiller2 — structural-similarity splits

`validation_structure={"group_on": "<column>"}` keeps whole groups on one side of every split, so a scaffold/cluster id gives you leakage-free grouped CV. Unlike the old `groups` parameter it supports repeated bagging, a group-aware non-bagged holdout, `group_time_on` for data that is both grouped and temporal, and `size_validation_on_groups` to size the validation method by group count rather than row count. (`groups` is now implemented on top of it and deprecated for removal in 2.0.)

## @falrrema — providing partitions directly, and purging

Explicit partitions work today, including from any sklearn splitter:

```python
splits = list(TimeSeriesSplit(n_splits=5).split(X, y))
predictor.fit(..., num_bag_folds=5, ag_args_ensemble={"custom_splits": splits})
```

**Purging/embargo is not supported** — there is no gap parameter between the train and validation windows, so a walk-forward split with purging still has to be constructed by hand and passed through `custom_splits`. That is a real gap and a narrower feature than this issue; please open a focused issue for it if you would like it tracked, and link back here.

---

Closing as resolved for the requested capability. #5628 proposed a `cv_splitter` parameter for this and has been closed as superseded — it predated `validation_structure`, and a `TabularPredictor` parameter that only reached bagging would have left the dynamic-stacking holdout and the non-bagged holdout splitting randomly, which `validation_structure` makes temporally forward as well.
